csodiaq/csodiaqParser.py

```diff
@@ -0,0 +1,329 @@
+        default=0.75,
+        help="When setting bin values, this option indicates how close an m/z value must be to the bin value. Default is 0.75.\nOptional.\nNOTE: Multiple targeted m/z values may fall within a range that a mass spectrometer can identify in one scan. Thus, m/z values are binned to prevent redundant reanalysis.\nExample: let's say we have the m/z values of 199.5 and 200.5, a binValueProximity value of 0.75, and a bin value of 200.0.\nBoth of these m/z values would be in the same bin, as they are both with 0.75 of 200.0.",
+    )
+
+
+def check_for_conflicting_args(args):
+    if args["command"] == "id" and args["histogram"] and args["noCorrection"]:
+        raise argparse.ArgumentTypeError(
+            "The histogram flag is invalidated by the noCorrection flag. Please inspect your input and remove one of the tags."
+        )
+    if args["command"] == "id" and args["correctionDegree"] and args["noCorrection"]:
+        raise argparse.ArgumentTypeError(
+            "The correctionDegree parameter is invalidated by the noCorrection flag. Please inspect your input and remove one of them."
+        )
+    if (
+        args["command"] == "targetedReanalysis"
+        and args["protein"]
+        and len(args["input"]["protein"]) == 0
+    ):
+        raise argparse.ArgumentTypeError(
+            "The protein argument requires the presence of protein FDR files to function. Please run the protein scoring workflow or remove the protein argument from your commands."
+        )
+
+
+def create_output_name(commandName):
+    return f"csodiaq-{commandName}-{time.strftime('%Y%m%d-%H%M%S')}"
+
+
+def create_new_output_directory_path(newDirectoryLocation, commandName):
+    if newDirectoryLocation[-1] == "/":
+        newDirectoryLocation = newDirectoryLocation[:-1]
+    newDirectoryName = create_output_name(commandName)
+    if os.path.isdir(newDirectoryLocation):
+        newDirectoryPath = os.path.join(newDirectoryLocation, newDirectoryName)
+    else:
+        newDirectoryHeader = newDirectoryLocation.split("/")[-1]
+        newDirectoryParentDirectory = "/".join(newDirectoryLocation.split("/")[:-1])
+        newDirectoryPath = os.path.join(
+            newDirectoryParentDirectory, f"{newDirectoryHeader}-{newDirectoryName}"
+        )
+    return newDirectoryPath
+
+
+class OutputDirectory:
+    def __init__(self, commandName):
+        self.commandName = commandName
+
+    def __call__(self, newDirectoryLocation):
+        if os.path.isfile(newDirectoryLocation):
+            raise argparse.ArgumentTypeError(
+                "The -o or --output argument must be a directory or to-be-created directory header, not an existing file."
+            )
+        if not os.path.isdir(os.path.dirname(newDirectoryLocation)):
+            raise argparse.ArgumentTypeError(
+                "The -o or --output argument directory requires an existing parent directory."
+            )
```

Comment:
`OutputDirectory` shouldn't have knowledge of the argument list; it just has `commandName` and `newDirectoryLocation` when called. I would remove the `ArgumentTypeError` and replace it with `ValueError`, indicating that the directory already exists. The parser should deal with raised exceptions as appropriate (if you want to report `ArgumentTypeError` to the user, then have the parser catch the ValueError and report.

Reply:
Alternatively, use the leading underscore to indicate that the class is not meant to be used outside of this context.

Reply:
I'm a little confused about the issue here. ArgumentTypeError is an argparse-specific error, and using it returns the provided message to the end user, indicating why the parsing failed. Raising a ValueError in an argparse.parse_args() call fails to show the error message, and instead says that the class object failed generally, as below:

### with ArgumentTypeError:
```
(venv) (base) cranneyc@awlpdcg556mm unit % csodiaq id -i <path>/DELETE.mzXML -l <path>/library_files -o <path>/outputDir -hist
usage: csodiaq id [-h] -o OUTPUT -i INPUT -l LIBRARY [-t MATCHTOLERANCE] [-nc] [-c CORRECTIONDEGREE] [-hist]
csodiaq id: error: argument -l/--library: The -l or --library argument must be an existing file (and not a directory).
```

### with ValueError:
```
(venv) (base) cranneyc@awlpdcg556mm unit % csodiaq id -i <path>/DELETE.mzXML -l <path>/library_files -o <path>/outputDir
usage: csodiaq id [-h] -o OUTPUT -i INPUT -l LIBRARY [-t MATCHTOLERANCE] [-nc] [-c CORRECTIONDEGREE] [-hist]
csodiaq id: error: argument -l/--library: invalid <csodiaq.csodiaqParser.LibraryFile object at 0x140e18bb0> value: '<path>/library_files'
```

If asking about the use of these classes generally, they are also used by the GUI to evaluate which fields were filled out incorrectly, highlighting offending ones as red and passing ones as green.

Am I misunderstanding the issue?

Reply:
```Python
import os
from csodiaqParser import OutputDirectory
out_dir = OutputDirectory("command")
os.mkdir("conflicting_dir")

out_dir("conflicting_dir")
```
That should raises ArgumentTypeError, informing the user that they used a wrong value for the -o argument, even though the command line was never used and argparse was not called. The error message is giving the wrong information about what went wrong; it's not that the `-o` parameter is wrong, it's that the file/directory already exists.
You can keep it the error as-is, but you should indicate that the class shouldn't be used outside of its current context (leading underscore `_OutputDirectory` vs `OutputDirectory`). Similarly for the others.

Reply:
Got it - it's purely about specifying to the end user that this is a class that should only be used in an argparse setting. Sounds good, I'll add underscores to reflect that. Thanks!